src/language-js/printer-estree.js

```diff
@@ -1808,7 +1808,11 @@ function printPathNoParens(path, options, print, args) {
         let res;
         if (isStringLiteral(n.value)) {
           const value = rawText(n.value);
-          res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
+          if (options.singleQuote === "all") {
+            res = "'" + value.slice(1, -1).replace(/'/g, "&#39;") + "'";
+          } else {
+            res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
```

Comment:
Would it be possible for JSX strings to have the same behavior as normal strings? (i.e. `"Let’s go!"` would always use double quotes regardless of the setting, and `'Let"s go!'` would similarly always use single)

Reply:
hmm I think this should only apply if `singleQuote: 'all'`. The assumption would be that if you don't explicitly opt in to using single-quotes in jsx, you'll always get double-quotes, no exceptions. thoughts?

Reply:
IMO `foo="bar"` and `foo={"bar"}` should get the same quotes (except when `singleQuote` is `js` and there’s a tie between single and double quotes).

Reply:
that is what's happening now. For the following:

```jsx
    <div
        single='foo'
        single2={'foo'}
        double="bar"
        double2={"bar"}
        singleDouble='"'
        doubleSingle="'"
        singleEscaped={'\''}
        doubleEscaped={"\""}
        singleBothEscaped={'\'"'}
        doubleBothEscaped={"\"'"}
    />
```

it gets converted to, when `singleQuote: all`:

```jsx
    <div
        single='foo'
        single2={'foo'}
        double='bar'
        double2={'bar'}
        singleDouble='"'
        doubleSingle="'"
        singleEscaped={"'"}
        doubleEscaped={'"'}
        singleBothEscaped={'\'"'}
        doubleBothEscaped={'"\''}
    />
```

and when `singleQuote: js`:

```jsx
    <div
        single="foo"
        single2={'foo'}
        double="bar"
        double2={'bar'}
        singleDouble="&quot;"
        doubleSingle="'"
        singleEscaped={"'"}
        doubleEscaped={'"'}
        singleBothEscaped={'\'"'}
        doubleBothEscaped={'"\''}
    />
```

Reply:
How about the string `'""\''`?

Reply:
IMO it should be the same.

Reply:
I'm happy to change it, [however it does seem like a deliberate decision in `printString`](https://github.com/prettier/prettier/blob/master/src/common/util.js#L493) to minimize the number of escapes by counting the types of quotes in each, and effectively disregarding the `singleQuote` option. In other words:

`'""\''` - will always be single-quote, regardless of options, since it has 2 double and 1 single inside
`'"\'\''` - will always be double-quote, regardless of options, since it has 1 double and 2 single inside

In order to force this to be "always the same" I'd basically have to bypass or remove that logic. 
I have no qualms doing so, but just wanted to double-check that that's the desired approach

Reply:
Sorry if I’m not communicating clearly :/

What I meant by “always be the same” is to use the `printString` algorithm for JSX, too.

Reply:
oooh, ok, yeah sorry I thought you meant the other way around, makes sense now, will do

Reply:
I agree, if escapes should be needed, unless it's like only one, it's much better to just convert it to a string expression instead of keeping it as a raw string with a bunch of HTML escapes. The HTML escapes also only make sense for ReactDOM; they likely won't do the right thing on react native.

Reply:
I pushed an update that uses the JS print function if any escape is needed, bypassing the "convert to html entities" escape method altogether. 

----

I did end up going into a few rabbit holes trying to get `foo="bar"` and `foo={"bar"}` to sync up which I might need to some help on [mentioned above](https://github.com/prettier/prettier/pull/4798#discussion_r199832547). 
The core issue is that `print` needs to now behave differently depending on what context it's in. i.e. `printString` needs to:
- use single quotes when singleQuote: all
- use single quotes when singleQuote: js **and** it's not a JSX attribute
- use double quotes when singleQuote: js **and** it is a JSX attribute 

The problem I'm having with the above is that the expression inside `{"bar"}` can be any JS expression, that sometimes will be a string literal, sometimes a concatenation of string literals and most times any other random expression. I tried going down the path of passing an extra contextual option to `print` when it's called from the JSX context, but that just ends up converting all strings to double-quotes inside the entire expression

TL; DR: when singleQuote: js, I cannot see how to sync `foo="bar"` with `foo={"bar"}` without also having things like `foo={() => { const a = "bar"; }}` also use double-quotes.